### ISSUE TITLE
Unable to create code version to development

### ISSUE THREAD

**Issue by bhunter** [jenkinsci/osf-builder-suite-for-sfcc-deploy-plugin#6]
Thank you for building and maintaining this!!

I've got the webDav and OCAPI data settings using client id "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" and set up the open api credentials in Jenkins to the same. The build however errors out with a 411 failed to create code version -see clip of console out below. It is successful when I build to my sandbox and in the API Explorer.
I can also use a client key created for Staging and Dev, w/ webdav and OCAPI settings that was made for develop and that throws the same error in the console below. It is also successful from the API Explorer and Postman.

-----Console output ----------
```
[+] Creating new code version
 - development-<redacted>-realm-corp.demandware.net (b26_20210325_dev)
Deleting 1 temporary files
ERROR: Failed to create the new code version! 411 - Length Required!
Response=<HTML><HEAD>
<TITLE>Bad Request</TITLE>
</HEAD><BODY>
<H1>Bad Request</H1>
Your browser sent a request that this server could not understand.<P>
Reference&#32;&#35;7&#46;4d012417&#46;1616693934&#46;35d99813
</BODY>
</HTML>
```

**Comment by danechitoaie** [comment]
We have few builds being done on development instances as well and didn't noticed this error. 
Last build was just done about 50 minutes ago and it worked:
```
[+] Creating new code version
 - development-eu01-something.demandware.net (b181_20210326_development)
 + Ok
```

Are you connecting directly to the SFCC development instance? or trough some kind of proxy? Maybe the proxy is removing the `Content-Length` header? As the 411 http error is related to this header missing: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/411

I just did a quick test and pointed the plugin to attempt to make the deployment to Ngrok endpoint instead of SFCC endpoint so I can capture the request and here it is bellow:
```
PUT /s/-/dw/data/v20_10/code_versions/b1_20210326_dev?client_id=abc HTTP/1.1
Host: 5b548fceff41.ngrok.io
User-Agent: Jenkins (OSF Builder Suite For Salesforce Commerce Cloud)
Content-Length: 0
Accept-Encoding: gzip
Authorization: Bearer abcd
```

The `Content-Length: 0` is set and sent in the request.

So maybe there's that HTTP proxy that I was asking about above that your requests are going trough and that proxy is removing this header?

**Comment by bhunter** [comment]
Ngrok is a good idea! 
I believe we do go through a proxy and will check in with our Network people to see if they have any info. It's odd though since the same Jenkins instance also deploys to our Staging instance without issue which is why I did not go there first.

**Comment by danechitoaie** [comment]
From what I know development has the eCDN (Cloudflare) on top, while staging does not. So maybe this error is triggered by Cloudflare (on dev - WAF?) while on Staging since you connect directly to SFCC instance they are not that strict (no WAF) and allow the request without the header as well.

**Comment by bhunter** [comment]
As it turns out I was wrong and we do not go through a proxy and our network does not mess with any headers. I've confirmed that development is the issue because we can create code versions on Staging (w/2FA) and Sandboxes (using client key and 'aaa...'). 

So if this is an issue on Cloudflare then how do I resolve it?

**Comment by bhunter** [comment]
Actually I CAN successfully hit the PUT code_versions/ endpoint using Postman! Is there a better way to view the headers than ngrok?

**Comment by bhunter** [comment]
Sorry to be a pain but I found that ngrok reports Content-Length:0 whether it is there or not. I sent a request using Postman with Content-Length unchecked. Ngrok reported that the Content-Length: 0 is there. Note the same Postman token header.

![image](https://user-images.githubusercontent.com/53947/112727014-bda8fe00-8ef6-11eb-8a19-6e8758db4421.png)
 

![image](https://user-images.githubusercontent.com/53947/112727072-f6e16e00-8ef6-11eb-9f29-309b327a6b3e.png)

**Comment by danechitoaie** [comment]
> Sorry to be a pain

No worries, I'm happy to help troubleshoot this.

> I found that ngrok reports Content-Length:0 whether it is there or not.

Hmm, then I'll have to think of a different way to test the headers. Any idea what else to use besides Ngrok? Ngrok was my go to tool for debugging http requests. 

Can you try to make same request with curl from same server where Jenkins is running? You can even do it from Jenkins. Create new job and run sh command in it.

Strange that we don't have this issue on development instances that we use. If I could reproduce this it would be easier to debug.

**Comment by danechitoaie** [comment]
Also are you on latest version of the plugin? Or an older one? 
Latest one is 1.1.4

**Comment by danechitoaie** [comment]
I've pushed a new version https://github.com/jenkinsci/osf-builder-suite-for-sfcc-deploy-plugin/commit/04db8c621af715b536083edd6277b4f02caf133d

where I'm setting an empty body `""` for the request. Maybe this helps forcing the Apache HTTP Client (the library I use for HTTP requests) to set the Content-Length header.

Can you try with this one and see if it works?

**Comment by bhunter** [comment]
Hey I am using your latest on my laptop to test. Now getting: 
```
Deleting 1 temporary files
ERROR: Failed to create the new code version! 415 - Unsupported Media Type!
Response={"_v":"19.8","fault":{"arguments":{"contentType":"text/plain; charset=UTF-8"},"type":"UnsupportedContentTypeException","message":"The media type 'text/plain; charset=UTF-8' in the Content-Type header isn't supported. Note: Only the UTF-8 charset is supported."}}
```

I tried to add the Content-Length header manually and that seemed to work. I can put in a PR.

**Comment by danechitoaie** [comment]
Merged and pushed a new version v1.1.6 
Thank you.

**Comment by bhunter** [comment]
Thank you we are using this for Development and Staging now!
